docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,492 @@
+* `persistence_ready` will wait for the system databases to be started.
+* `ready_to_run` will wait for all of the above.
+
+With a few exceptions,
+the majority of EdgeX services will block on `ready_to_run`'s HEALTHCHECK.
+
+### Seeding of initial secrets
+
+For "mass production" of EdgeX appliances,
+it is not feasible to manually touch and configure each appliance.
+It is desirable to inject certain secrets directly into the bootstrapping flow.
+These secrets include:
+
+* TBD
+* TBD
+* TBD
+
+After secret store initialization,
+a JSON file containing seed secrets will be processed and injected into the KV secrets engine.
+This JSON file will be of the format:
+
+<pre>
+{
+  "path/to/secret": {
+    anything
+  }, ...
+}
+</pre>
+
+This file will be fetched from a preconfigured URL (file://, https://) passed to the bootstrapping container.
+
+
+### Bootstrapping flow
+
+#### install
+
+This ADR proposes that Vault be changed
+from using the Consul storage backend
+to using the raft storage backend.
+As both storage backends are HA-capable,
+there is no functional impact to this change.
+However, this change greatly simplifies the installation phase,
+as Vault can be initialized first
+and then generate bootstrapping secrets needed by Consul.
+Without this change,
+Consul must be stopped and started multiple times,
+as Consul would be needed by Vault
+to generate Consul's bootstrapping secrets.
+This change also eliminates the need
+to generate an ACL token
+for Vault to use Consul's storage backend.
+
+The purpose of the install flow is to generate EdgeX bootstrapping secrets
+defined in
+[Summary of bootstrapping secrets](#Summary%20of%20bootstrapping%20secrets).
+
+The rough flow of events is as follows:
+
+1. Skip to the end if bootstrapping has already been done.
+
+Vault bootstrapping:
+
+1. Bring up vault on localhost with TLS disabled (bootstrapping configuration)
+1. Initialize a blank Vault and immediately unseal it
+1. Encrypt the Vault keyshares and revoke the root token
+1. Generate a new root from the keyshares
+1. Generate an on-device root CA (see https://learn.hashicorp.com/vault/secrets-management/sm-pki-engine)
+1. Create an intermediate CA for TLS server authentication
+1. Sign the intermediate CA using the root CA
+1. Configure policy for intermediate CA
+1. Generate and store leaf certificates for
+   Consul,
+   Vault,
+   PostgreSQL,
+   Kong (admin),
+   Kong (external),
+   Redis (v6),
+   Mosquitto
+1. Deploy the PKI to the respective services' secrets area
+1. Write the production Vault configuration (TLS-enabled)
+
+Consul bootstrapping:
+
+1. Write the production Consul configuration (TLS-enabled)
+   * HTTPS on port 8501
+   * Disable HTTP on 8500
+   * Disable DNS
+   * Specify ca_file, cert_file, and key_file
+1. Generate a Serf encryption key and save it to Vault
+1. Start consul and generate a management token
+1. Configure the Consul secrets engine in Vault
```

Comment:
I thought these go away with `Raft`

Reply:
No.  Consul secrets engine allows use of a Vault API to create Consul authentication tokens.  This puts Vault in the position of creating and managing secrets, versus having to talk to Vault to create vault tokens and Consul to create consul tokens.